### PR TITLE
PEP 757: Use union

### PR DESCRIPTION
**Pull request by vstinner**
----
📚 Documentation preview 📚: https://pep-previews--4111.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
@skirpichev: What do you think of using an union in the export?

**Review by skirpichev — COMMENTED** [review]
With current proposal we don't gain anything.  E.g. some contract like "this will not fail for integers".

I'm neutral on keeping the ``kind`` field wrt using the return value of PyLong_Export() to encode the union type (yet the struct, probably, will be bigger), as https://github.com/python/peps/pull/4026/commits/c0a725b64d9e6253309fe41c0afafed0bafff870 does.

**Comment by vstinner** [comment]
I'm not changing any contract. It's just the structure which changes to make it more obvious which members should be used depending on the kind.

File: peps/pep-0757.rst [hunk 1]
```diff
@@ -270,16 +284,18 @@ Code::
...
 
-        if (long_export.digits) {
-            mpz_import(z, long_export.ndigits, int_digits_order, int_digit_size,
-                       int_endianness, int_nails, long_export.digits);
-            if (long_export.negative) {
+        if (long_export.kind == PyLongExport_ARRAY_KIND) {
+            mpz_import(z, long_export.data.array.ndigits,
+                       int_digits_order, int_digit_size, int_endianness,
+                       int_nails, long_export.data.array.digits);
+            if (long_export.data.array.negative) {
```
Comment (NameNoQuality) [inline peps/pep-0757.rst:291]: #4113

**Review by NameNoQuality — APPROVED** [review]
#4113

File: peps/pep-0757.rst [hunk 2]
```diff
@@ -107,37 +107,51 @@ Export API
...
-      The native integer value of the exported :class:`int` object.
-      Only valid if :c:member:`digits` is ``NULL``.
+      .. c:namespace:: NULL
 
-   .. c:member:: uint8_t negative
+      ========================================  =====
+      Kind                                      Value
+      ========================================  =====
+      .. c:macro:: PyLongExport_VALUE_KIND      ``1``
+      .. c:macro:: PyLongExport_ARRAY_KIND      ``2``
```
Comment (skirpichev) [inline peps/pep-0757.rst:127]: Should we expose these details?  Lets just keep symbolic constants.

**Review by skirpichev — COMMENTED** [review]
> I'm not changing any contract.

Bad news.  I would buy such change if one allows us to skip error checking.  Or simplify current implementation (e.g. if only one export type will be used).

> It's just the structure which changes to make it more obvious which members should be used depending on the kind.

I think that current PEP shape is simple enough.  Lets not rewrite it again ;)  I doubt that in future we will need something else than current two versions of export.

With union the structure will be, probably, more compact.  This is only benefit of this change, for me.  Not sure if it worth.

Count me neutral on this.

**Comment by vstinner** [comment]
Ok, I close this PR.

**Pull request closed without merging**